Implement '/gameTitles/createTitle' route

Ahuanmarca opened the pull request:
**Creates new gameTitle and a relation with one or more genres**
- Insert into 'gameTitles' collection IF NOT EXISTS
- Insert into 'genres' collection IF NOT EXISTS (supports multiple genres)
- Insert into 'genres_gameTitles' collection IF NOT EXISTS (supports multiple relations)
- Returns object containing only newly created / stored data

Test it with Postman!

Route:
http://localhost:3000/gameTitles/createTitle

Body (raw):

```
{
    "title": "Test_Title_1",
    "description": "This description is a test",
    "image": "https://cdn.mobygames.com/covers/3906896-final-fantasy-iii-snes-front-cover.jpg",
    "genres": ["test_genre_1", "test_genre_2", "adventure"]
}

```

Ahuanmarca committed:
Implement '/gameTitles/createTitle' route: Creates new gameTitle and a relation with one genre.

How it works:
- Attempts to create a document in gameTitles collection, then a document in genres_gameTitles collection to create a relation between the gameTitle and a genre.
- It will create the gameTitle only if not exists.
- It will create the relation with a genre only if not exists.
- If the genre does not exist, will create it.

Main function:
- createTitle(), inside gameTitles.service.js. Takes data from form: title, description, image, genre.

The createTitle() function calls several functions:
  - gameTitlesRepository.getByTitle(title)
  - gameTitlesRepository.createTitle({ title, description, image })
  - genresRepository.getGenreByName(genre)
  - genresRepository.createGenre(genre)
  - genres_gameTitlesRepository.findByGenreAndTitle(genreId, newTitle._id)
  - genres_gameTitlesRepository.createGameTitleGenreRelation(newTitle._id, genreId)

Implemented a model and a repository for the genres_gameTitles relation:
- genres_gameTitles.js: Model for a table with the relation between game titles and genres (a join table)
- genres_gameTitles.js implements 3 new functions that are needed to create a gameTitle with relation to genres
  - getAllGenreTitleRelations()
  - createGameTitleGenreRelation(gameTitleId, genreId)
  - findByGenreAndTitle(genreId, titleId)

Additional functions created:
Some additional functions were created so createTitle() works ok.
- Inside gameTitles.repository.js:
  - getByTitle() - Created because we need to check if a gameTitle already exists when the user is trying to create a new one.
  - createTitle() - Creates the title only in the gameTitles collection
- Inside genres.repository.js
  - getGenreByName() - We need to check if a genre already exists when the user tries to create a title (with a genre)
  - createGenre() - If the user creates a title with a genre that doen't exist, we can create it

nuinn commented on src/api/gameTitles/gameTitles.service.js:
```diff
@@ -17,8 +26,48 @@ async function getByProductId(productId) {
+  let newTitle;
+  const foundGameTitle = await gameTitlesRepository.getByTitle(title);
+  if (foundGameTitle) {
+    newTitle = foundGameTitle;
```
The function is named "create" and this variable is called "newTitle", but neither describes what is happening in this instance.
Would it not make more sense to notify the user that their attempt to add a title has not been successful as it already exists in the DB, rather than returning the same information in both successful and unsuccessful scenarios?

Ahuanmarca replied:
Yes, you are right! I'm aware of this problem. Do you think it's enough to just notify the user (in the response, somehow) when the database was not modified?

Ahuanmarca committed:
Implement: When creating gameTitle, response tells wether new informatioin was created in the database

nuinn commented on src/api/gameTitles/gameTitles.service.js:
```diff
@@ -17,8 +26,57 @@ async function getByProductId(productId) {
 
+async function createTitle(newTitleData) {
+  const {
+    title, description, image, genre,
```
What type of data is "genre"? What happens if the gameTitle has multiple genres?

Ahuanmarca replied:
Good questions!
The route only accepts one genre, but a lot of titles have multiple genres, so we need to make the route capable of that.
Currently, genre is a string (comming from a form on the front-end). I think it should be an array of strings.
I'll work on it.

Ahuanmarca committed:
Implement creating multiple genres and genre_title relations when creating a gameTitle

Almost completely refactored createTitle on gameTitles.service.js to support async iteration!

Ahuanmarca committed:
Merge branch 'develop' into createGameTitle

nuinn commented on src/api/gameTitles/gameTitles.repository.js:
```diff
@@ -1,22 +1,35 @@
-import gameTitleModel from './gameTitles.model.js';
```
I understand foregoing CamelCase with the genres_gameTitles router, but I'm not sure why you've done it here

Ahuanmarca replied:
Eslint warns that "constructor" functions should be named in UpperCamelCase.  The models are constructor functions, but eslint warning is triggered only when you use the 'new' keyword to invoke the function...
Sadly we have several constructor functions in camelCase instead of UpperCamelCase, I've corrected some of them but not all.

Ahuanmarca committed:
Merge branch 'develop' into createGameTitle.
Merges an updates upstream into a topic branch.

andreamir commented on src/api/genres/genres.repository.js:
```diff
@@ -10,7 +10,27 @@ async function getGenreById(id) {
+  return newGenre;
+}
+
+async function createManyGenres(genres) {
```
👍🏻

andreamir commented on src/api/gameTitles/gameTitles.service.js:
```diff
@@ -17,8 +26,83 @@ async function getByProductId(productId) {
 }
 
+/**
+ * createTitle implementation:
```
We shouldn't merge comments to main

andreamir commented on src/api/gameTitles/gameTitles.service.js:
```diff
@@ -17,8 +26,83 @@ async function getByProductId(productId) {
+    .filter((genre) => !genre.found)
+    .map((genre) => ({ name: genre.genre }));
+  const newGenres = await genresRepository.createManyGenres(genresToCreate);
+  // created.genres.push(...newGenres); // ! Throws Error, why?
```
The line, which is commented, but it also doesn't work?

Ahuanmarca committed:
Cleaned comments

Ahuanmarca committed:
Merge branch 'develop' into createGameTitle
Merges an updates upstream into a topic branch.

JonaTheBridge commented on src/api/gameTitles/gameTitles.service.js:
```diff
@@ -17,8 +20,69 @@ async function getByProductId(productId) {
+    .filter((genre) => !genre.found)
+    .map((genre) => ({ name: genre.genre }));
+  const newGenres = await genresRepository.createManyGenres(genresToCreate);
+  newGenres.forEach((newGenre) => created.genres.push(newGenre));
```
esta linea no puede simplificarse haciendo created.genres = newGenres???

Ahuanmarca replied:
Propongo simplificarla así para "llenar" el array en vez de reasignarlo:
`created.genres.push(...newGenres);`

JonaTheBridge commented on src/api/gameTitles/gameTitles.service.js:
```diff
@@ -17,8 +20,69 @@ async function getByProductId(productId) {
+  const newGenres = await genresRepository.createManyGenres(genresToCreate);
+  newGenres.forEach((newGenre) => created.genres.push(newGenre));
+
+  const genrePromises = genres.map(async (genre) => {
```
esto lo haces en la linea 43 no?

Ahuanmarca replied:
En la línea 43 se buscan 'genres' para crear los que no existen (es decir, no necesariamente vamos a **encontrar** todos los que estamos buscando.
¿Por qué repito la búsqueda? Luego de (posiblemente) crear los que no existían, hago un nuevo query a la base de datos porque necesito todos los genres para vincularlos con el gameTitle (neceisto sus IDs).

JonaTheBridge commented on src/api/gameTitles/gameTitles.service.js:
```diff
@@ -17,8 +20,69 @@ async function getByProductId(productId) {
+    const foundGenre = await genresRepository.getGenreByName(genre);
+    return foundGenre;
+  });
+  const genresToRelate = await Promise.all(genrePromises);
```
y esto en la 47

Ahuanmarca replied:
Esta búsqueda repetida es por el mismo motivo que la anterior. Posiblemente hay una mejor manera de hacerlo 😅

JonaTheBridge commented on src/api/gameTitles/gameTitles.service.js:
```diff
@@ -17,8 +20,69 @@ async function getByProductId(productId) {
+      found: !!foundTitleGenreRelation,
+    };
+  });
+  const titleGenreRelationUpsertMap = await Promise.all(titleGenreRelationUpsertPromises);
```
MUY buen uso de Promise.all durante todo el archivo, enhorabuena!

Ahuanmarca replied:
Gracias!

JonaTheBridge commented on src/api/gameTitles/gameTitles.service.js:
```diff
@@ -17,8 +20,69 @@ async function getByProductId(productId) {
   return gameTitles;
 }
 
+async function createTitle(newTitleData) {
```
aquí me sobra la palabra Title. Ya estamos en **gameTitles**.service.js, simplemente con create ya sabemos qué vamos a crear

JonaTheBridge commented on src/api/genres/genres.repository.js:
```diff
@@ -10,7 +10,27 @@ async function getGenreById(id) {
   getAll,
+  getGenreByName,
+  createGenre,
+  createManyGenres,
```
en estas 3 últimas lineas veo de nuevo el problema de poner qué vamos a crear o a obtener. Para mi tiene más sentido que las funciones se llamen getByName, create y createMany. De esta manera se usarían escribiendo **genres**Repo.create() y no **genres**Repo.create**Genre**()

JonaTheBridge commented on src/api/genres_gameTitles/genres_gameTitles.repository.js:
```diff
@@ -0,0 +1,34 @@
+  createGameTitleGenreRelation,
+  getAllGenreTitleRelations,
+  findByGenreAndTitle,
+  createManyGenreAndTitleRelations,
```
y aquí lo mismo, fíjate las incoherencias que pueden llegar a cometerse: createGame**TitleGenre**Relation, getAll**GenreTitle**Relations, findBy**GenreAndTitle**, createMany**GenreAndTitle**Relations

Si evitamos poner en los nombres de las funciones lo que ya está en el nombre del fichero (**genres_gameTitles**.repository.js) no cometeremos este tipo de incoherencias

Ahuanmarca replied:
He corregido todas menos `findByGenreAndTitle` porque también sería posible encontrar únicamente por Genre o por Title, pero este method intenta encontrar por ambos  atributos.

Ahuanmarca committed:
Improved method naming, improved mongoose queries, simplified gameTitle service

Ahuanmarca committed:
Cleaned some unused methods

Ahuanmarca committed:
Adapted to develop to minimize conflicts, in preparation to merge

Ahuanmarca committed:
Merge branch 'develop' into createGameTitle

Ahuanmarca committed:
Uncommented isLogged middleware
